appcues/src/test/java/com/appcues/analytics/ExperienceLifecycleTrackerTest.kt

```diff
@@ -0,0 +1,188 @@
+        val initialState = RenderingStep(experience, 0, true)
+        val action = EndExperience(destroyed = false, markComplete = false)
+        val scope = initScope(initialState)
+        val stateMachine: StateMachine = scope.get()
+        val analyticsTracker: AnalyticsTracker = scope.get()
+
+        // WHEN
+        stateMachine.handleAction(action)
+
+        // THEN
+        verify(exactly = 0) { analyticsTracker.track("appcues:v2:step_completed", any(), any(), any()) }
+    }
+
+    @Test
+    fun `RenderingStep SHOULD track step_completed WHEN action is EndExperience markComplete=false on last step`() = runTest {
+        // GIVEN
+        val experience = mockExperience()
+        val initialState = RenderingStep(experience, 3, false)
+        val action = EndExperience(destroyed = false, markComplete = false)
+        val scope = initScope(initialState)
+        val stateMachine: StateMachine = scope.get()
+        val analyticsTracker: AnalyticsTracker = scope.get()
+
+        // WHEN
+        stateMachine.handleAction(action)
+
+        // THEN
+        verify { analyticsTracker.track("appcues:v2:step_completed", any(), any(), any()) }
+    }
+
+    @Test
+    fun `RenderingStep SHOULD track step_completed WHEN action is EndExperience markComplete=true AND not on last step`() = runTest {
+        // GIVEN
+        val experience = mockExperience()
+        val initialState = RenderingStep(experience, 0, true)
+        val action = EndExperience(destroyed = false, markComplete = true)
+        val scope = initScope(initialState)
+        val stateMachine: StateMachine = scope.get()
+        val analyticsTracker: AnalyticsTracker = scope.get()
+
+        // WHEN
+        stateMachine.handleAction(action)
+
+        // THEN
+        verify { analyticsTracker.track("appcues:v2:step_completed", any(), any(), any()) }
+    }
+
+    // Helpers
+    private suspend fun initScope(state: State): Scope {
+        val scopeId = UUID.randomUUID().toString()
+        val app = createKoinApp(scopeId, state)
+        val scope = app.koin.getScope(scopeId)
+        val machine: StateMachine = scope.get()
+        val coroutineScope: AppcuesCoroutineScope = scope.get()
+
+        coroutineScope.launch {
+            // this collect on the stateFlow simulates the function of the UI
+            // that is required to progress the state machine forward on UI present/dismiss
+            machine.stateFlow.collect {
+                when (it) {
+                    is BeginningStep -> {
+                        it.presentationComplete.invoke()
+                    }
+                    is EndingStep -> {
+                        it.dismissAndContinue?.invoke()
+                    }
+                    // ignore other state changes
+                    else -> Unit
+                }
+            }
+        }
+
+        coroutineScope.launch {
+            scope.get<ExperienceLifecycleTracker>().start(UnconfinedTestDispatcher())
```

Comment:
I got why this was done, there is also another way we could do it. simply by creating an interface and having Koin give different implementations when its running the SDK vs when its testing.

```kotlin
interface AppcuesDispatchers {
   fun main()
   fun io()
}

// for real exe
class DefaultDispatchers: AppcuesDispatchers {
   fun main() = Dispatchers.MAIN
   fun io() = Dispatchers.IO
}

// for tests
class TestDispatchers: AppcuesDispatchers {
  val dispatcher = UnconfinedTestDispatcher()
  fun main() = dispatcher
  fun io() = dispatcher
}
```

if for some reason we see this again we can consider refactoring to this if it looks interesting

Reply:
sorry i meant to make a note about that and left it as a pending comment but glad you got it anyway.  Yes, that sounds like a good suggestion to keep in mind to use Koin to assist there, thanks!